exp/bufferpool/pool.go

```diff
@@ -4,49 +4,151 @@ package bufferpool
 import "sync"
 
 const (
-	nBuckets = 20
+	defaultMinSize     = 1024
+	defaultBucketCount = 10
 )
 
-// A default global pool.
-var Default Pool
+var defaultPool = &Pool{}
 
-// A pool of buffers, in variable sizes.
+// Set the global pool.  This can be used to override the default implementation
+// to optimize for different allocation profiles. For example, applications that
+// target embedded platforms may wish to lower the minimum buffer size from 1KiB
+// to 512B, or less.
+//
+// The default implementation uses BasicPool, which is suitable for a wide range
+// of applications.  Most applications will not benefit from using a custom pool,
+// so be sure to profile.
+//
+// Set panics if p == nil.
+func Set(p *Pool) {
+	if p == nil {
+		panic("Pool(nil)")
+	}
+
+	defaultPool = p
+}
+
+// Get calls .Get() on the global pool, and resizes the resulting buf
+// to size.
+func Get(size int) []byte {
+	return defaultPool.Get(size)[:size]
+}
+
+// Put calls .Put(buff) on the global pool after zeroing the supplied
+// slice. Note that only the first len(buf) bytes are zeroed, and not
+// the full backing array.
+//
+// TODO:  should we zero out the entire array?  Users may have resized
+// the slice, so it may contain sensitive data.
+func Put(buf []byte) {
+	for i := range buf {
+		buf[i] = 0
+	}
+
+	if capacity := cap(buf); capacity > 0 {
+		defaultPool.Put(buf[:capacity])
+	}
+}
+
+// Pool maintains a list of BucketCount buckets containing buffers
+// of exponentially-increasing size, starting from MinSize. Buffers
+// whose capacity exceeds MinSize << BucketCount are ignored and left
+// to the garbage-collector.
+//
+// The zero-value Pool is ready to use, defaulting to BucketCount=10
+// and MinSize=1024 (1KiB - ~1MiB buffer sizes).  Most applications
```

Comment:
Nit: the tilde is unnecessary; Mi is a power of two prefix, so this is actually exact (the base 10 prefixes don't have the i).

Reply:
Unless my arithmetic is mistaken (totally possible), the slices contained in the largest bucket will have a capacity of at most 1048575 Bytes (.99 MiB).  See `TryPut`:

```
func (b bucket) TryPut(buf []byte) (ok bool) {
	if ok = b.minCap <= cap(buf) && b.capLimit > cap(buf); ok {
		b.pool.Put(buf)
	}

	return
}
```

Reply:
Bucket i should have capacity `1024 << i` = `1024 * 2 ^ i` = `2 ^ 10 * 2 * i` = `2 ^ (10 + i)` so with 10 buckets the largest should be `2 ^ 20` = 1MiB. What am I missing?

Reply:
Actually, with 10 buckets the last bucket is `i = 9`, so the limit is actually half a MiB...